src/main/java/org/jitsi/srtp/crypto/OpenSslAesCipherSpi.java

```diff
@@ -0,0 +1,726 @@
+                    ("ECB mode cannot use IV");
+            }
+        }
+        else if (iv == null)
+        {
+            /* According to the SPI spec we should use random to generate the
+             * IV in this case if we're encrypting, but we never want to do this for SRTP.
+             */
+            throw new InvalidAlgorithmParameterException
+                ("IV parameter missing");
+        }
+
+        byte[] keyParam = null;
+        long cipherType = 0;
+        if (key != this.key)
+        {
+            this.key = key;
+            keyParam = key.getEncoded();
+            cipherType = getCipher(key);
+        }
+
+        if (newIvLen != 0)
+        {
+            if (!EVP_CipherSetIVLen(ctx, newIvLen))
+            {
+                throw new InvalidAlgorithmParameterException
+                    ("Unsupported IV length " + newIvLen);
+            }
+        }
+
+        if (!EVP_CipherInit(ctx, cipherType, keyParam, this.iv, openSslEncryptMode))
+        {
+            throw new InvalidKeyException("AES_CTR_CTX_init");
+        }
+    }
+
+    @Override
+    protected void engineInit(int opmode, Key key, AlgorithmParameters params,
+        SecureRandom random)
+        throws InvalidKeyException, InvalidAlgorithmParameterException
+    {
+        AlgorithmParameterSpec spec = null;
+        if (params != null)
+        {
+            try
+            {
+                if (cipherMode == CipherMode.GCM)
+                {
+                    spec = params.getParameterSpec(GCMParameterSpec.class);
+                }
+                else
+                {
+                    spec = params.getParameterSpec(IvParameterSpec.class);
+                }
+            }
+            catch (InvalidParameterSpecException e)
+            {
+                throw new InvalidAlgorithmParameterException(e);
+            }
+        }
+        engineInit(opmode, key, spec, random);
+    }
+
+    @Override
+    protected byte[] engineUpdate(byte[] input, int inputOffset, int inputLen)
+    {
+        int bufNeeded = getOutputSize(inputLen, false);
+        byte[] buf = new byte[bufNeeded];
+        try
+        {
+            int len = engineUpdate(input, inputOffset, inputLen, buf, 0);
+            assert(len == bufNeeded);
+        }
+        catch (ShortBufferException e)
+        {
+            /* Shouldn't happen, we allocated enough. */
+            throw new IllegalStateException(e);
+        }
+        return buf;
+    }
+
+    /**
+     * Call EVP_CipherUpdate, throwing on failure.
+     */
+    private void doCipherUpdate(byte[] input, int inputOffset, int inputLen,
+        byte[] output, int outputOffset)
+    {
+        if (!EVP_CipherUpdate(ctx, input, inputOffset, inputLen, output,
+            outputOffset))
+        {
+            throw new IllegalStateException("Failure in EVP_CipherUpdate");
+        }
+    }
+
+    private int doGcmDecryptUpdateWithBuffer(byte[] input,
```

Comment:
Is this whole buffering thing really necessary? For a general purpose OpenSSL SPI the answer is obviously yes, but we never call cipher.update in jitsi-srtp, it's always directly cipher.doFinal.

Reply:
I was thinking I should write a general purpose OpenSSL SPI, but as you point out we don't use it, and as such the code isn't ever tested, which is bad.  I'm considering having it throw `UnsupportedOperationException` instead.

Reply:
I've now done this (making your previous comments on this code moot, I believe.)